### ISSUE TITLE
Backend Refactor

### ISSUE THREAD

**Issue by ncilfone** [fidelity/spock#2]
Shifting away from dataclasses since the current implementation is kludge consisting of the current Python 3.7 dataclass implementation overridden to only allow kwarg defs and prevent an ordering requirement (which affects inheritance) of default values.

Possible Solutions:

* attrs - https://github.com/python-attrs/attrs

* pydantic - https://github.com/samuelcolvin/pydantic/

Leaning towards attrs as they provide a bit more functionality.

**Comment by ncilfone** [comment]
@sidnarayanan backend branch has a sketch for the shift to attrs. Shows how the interface might change.

**Comment by sidnarayanan** [comment]
Couple quick questions:

- This will change the API, right? Attributes will have to be defined as `attr.ib` instances via `katra`, as opposed to the current method of using type-hints 

- Why do we need to enumerate allowed classes via `SpockTypes`? Why not open it up to any class and let `attr.validators` handle type-checking? We can catch the exception if a validator doesn't exist for the provided type. I'm thinking this flexibility will be useful for #6. And also it allows the user to write:
```python
@spock 
Config:
    attribute = katra(int)
```
which feels more idomatic than
```python
@spock
Config:
    attribute = katra(SpockTypes.INT)
```

**Comment by ncilfone** [comment]
@sidnarayanan 

Responding in order:

- Yes the API will change with this however this way (using attrs.ib) does strict type checking and optional handling with validators when setting a value, which means we don't need any internal logic anymore which is nice (and much cleaner -- the code that is on 4b0996e is basically it... or so I think). Technically we could go back to type hinting (see [here](https://www.attrs.org/en/stable/types.html) but we lose a lot of the benefits of attrs (most of the advanced features)

- Would be fine for basic types like str and int but the challenge comes with list and tuple which are python generic types and then on top of that with subscripted types. For lists we would be ok because we can map from `list` to `List` within `typing` but for specifying something like `List[int]` we might have to do some magic... I think it's possible to go without the Enum but I'll have to think it through.


TLDR;
1 - Yes the API kinda has to change for attrs if we want the good stuff
2 - I think we can make it work if you like the leaner style...

**Comment by ncilfone** [comment]
@sidnarayanan Re 2nd point - from a practical aspect wouldn’t it be better to know before run time what you can and cannot define instead of waiting for an exception at run time telling you that the type isn’t supported... the exception would beg the question ‘what is supported’? An the Enum (with the type hint in the katra) is a simple way to do that (we could just have a list)

**Comment by sidnarayanan** [comment]
Point 1: Totally fine with changing the API. I wonder if, down the line, we can revert to type-hinting as an option, but use some introspection to translate to katra calls. Something like:

```python
@spock_dataclass 
class Config:
    param: int

def spock_dataclass(cls):
    for k,v in cls.__dataclass_fields__:
        typed = k.type # set by type hint
        attribute = katra(typed)
```

Point 2: I agree that it would be better to know at declaration time. But now that I think about it, doesn't `attr.validators.instance_of` support everything? I think this line:
```python
        x = attr.ib(validator=attr.validators.optional(attr.validators.instance_of(typed.value)), default=default)
```
could just be changed to:
```python
        x = attr.ib(validator=attr.validators.optional(attr.validators.instance_of(typed)), default=default)
```

**Comment by ncilfone** [comment]
@sidnarayanan 

8b3ec70 should handle a lot of the discussion above. Type hinting interface is back (relying on the user understanding `typing`) for the attr backend.

**Comment by sidnarayanan** [comment]
Was poking around the implementation, found probably unintended behavior in optional checking:
```python
In [14]: typed._handle_optional_typing(typing.Dict[int, None])
Out[14]: (int, True)
```
It sees the `None` arg. If this is indeed unintended, we can explicitly verify that's equal to a `Union` since this holds:
```python
In [19]: typing.Optional[int] == typing.Union[int, None]
Out[19]: True
```

**Comment by ncilfone** [comment]
@sidnarayanan 

f0edd3d should be a working split of the backend + new attrs backend with a common interface. Syntax doesn't change at all minus the attrs backend using Typing instead of the old definition of custom args. Dataclass backend passes all of the existing tests....

TODO: Create all tests for the new attrs backend. Create docs for the new backend (although syntax is very similar so shouldn't be much). Attrs does not have equivalent ChoiceTypeArg and SavePathArg. Still need a way to handle nested GenericAlias (e.g. List[List[int]]). Need to figure out how to alias the import of spock.backend.dataclass.args to spock.args so old code doesn't break.

**Comment by sidnarayanan** [comment]
@ncilfone 

Looks good to me! I like the use of `spock` vs `spock_config` to handle backwards compatibility and avoid name clashing.

Incidentally, this gets us halfway to covering #6. Added tests in https://github.com/sidnarayanan/spock/commit/e3fd09e191f936a99e371bfea4579a3025d99bf2 that show that it supports nested configuration in the case that defaults are provided. All that remains is adding the arg builder part of it.

**Comment by ncilfone** [comment]
#15 resolves
